Make IonizationTunnelEnvelopeAveraged inherit from IonizationTunnel

GianlucaOberreit opened the pull request:
- Make IonizationTunnelEnvelopeAveraged inherit from IonizationTunnel
- Make the KAG and TL BSI models inherit from IonizationTunnel
- Separate electron creation, ionization current computation and electric field calculation into separate methods
- Remove legacy code and unnecessary includes

TODO:
- Make tests for Envelope model more stringent for electron momentum initialisation

GianlucaOberreit committed:
Create TotalIonizPotNew and check whether the same result is obtained as with TotalIonizPot

GianlucaOberreit committed:
Move ionization current computation out of IonizationTunnel operator()

GianlucaOberreit committed:
Move new electron creation out of IonizationTunnel operator()

GianlucaOberreit committed:
make Epart more flexible and pull electric field calculation out of IonizationTunnel operator()

GianlucaOberreit committed:
make IonizationTunnelEnvelope inherit from IonizationTunnel

GianlucaOberreit committed:
fixes

GianlucaOberreit committed:
Move from templates to inheritance for IonizationTunnel and BSI. Correct bug in IOnizationTunnelEnvelopeAveraged. Remove unnecessary imports

GianlucaOberreit committed:
Update IonizationFactory with new setup

GianlucaOberreit committed:
Remove more unused headers

GianlucaOberreit committed:
correct bug in envelope

GianlucaOberreit committed:
Clean up. Remove some imports.

Some imports are replaced with forward declarations because apparently
this is good for compilation times.

GianlucaOberreit committed:
Pass ElectricFields as const reference. Make Z unsigned int everywhere

GianlucaOberreit committed:
handle E.abs=0

GianlucaOberreit committed:
Introduce SimulationContext. reduce inconsistency.

Introduce SimulationContext nested class with members Particle*, Patch*
and Proj* to simplify method calls. Reduce inconsistency in method
parameter order.

GianlucaOberreit commented:
Last commit has passed all 1D benchmarks for me.

GianlucaOberreit committed:
More cleanup

GianlucaOberreit committed:
More cleanup

JuanJo-Silva commented:
Hi, I'm working on porting the ionization module to GPU. It would be good to synchronize our developments so our modifications are compatible. If I understand correctly, are you working towards merging all ionization models into one class? 

I agree about splitting the computation of the current ionization and the creation of particles into different functions. These modifications isolate the Monte-Carlo routine, which is mostly common for all models. For the division of the different bsi models of the ionization tunnel into different files, I have another solution which is more compatible with the GPU porting of the other physical modules. It imply putting those restrictions into the ionization rate function.

GianlucaOberreit commented:
Hello,

> It would be good to synchronize our developments so our modifications are compatible.

Absolutely, I think I'm mostly done on my end, if everything seems to work correctly.

> If I understand correctly, are you working towards merging all ionization models into one class?

My aim is to consolidate all the logic common to the ionization models (except IonizationFromRate) into the IonizationTunnel class and have the other three (EnvelopeAveraged, KAG and TL) inherit from this.

> I agree about splitting the computation of the current ionization and the creation of particles into different functions. These modifications isolate the Monte-Carlo routine, which is mostly common for all models.

That's good to read!

> For the division of the different bsi models of the ionization tunnel into different files, I have another solution which is more compatible with the GPU porting of the other physical modules. It imply putting those restrictions into the ionization rate function.

So all the BSI would all be grouped into the single class IonizationTunnel? If so, I imagine the same restrictions would be applied to the constructor for the TL model?

GianlucaOberreit committed:
Correct bug in override for electron creation in envelope.

TODO: make tests more stringent

GianlucaOberreit committed:
Warnings cleanup. Add failsafe in envelope ionizationRate. Simplification.

GianlucaOberreit committed:
modify ionization class attribute names to obey conventions

GianlucaOberreit committed:
Make Epart const&

I'm on the const& train, no side effects
I'm on the const& train, strict semantics checks
I'm on the const& train
Stopping crashes and burns, with const& confirmed
I'm on the const& train, I love that stuff
I'm on the const& train and I can never get enough
I'm on the const& train, never to return, no

JuanJo-Silva commented:
I had just read your commits, nice work! It’s very nice that you already validated all your modifications so quickly. Are you planning to make the ionisation rate also inherit from the ionisation tunnel?

Regarding the BSI modules, in terms of computational operations, they’re very similar between them. They just differ in the correction terms that are calculated and used at the end of the ionization rate function. I was wondering if it would more easier to pass an argument to the ionization rate function to decide (through conditionals) which terms to calculate and apply. This could avoid a lot of code repetition and make the porting to GPU easier, but it’s unclear to me if it will affect CPU performance.  

Also, the constructor of the tunnel ionisation class allocates a lot of attribute vectors that are just arithmetic transformations from the ionization tables. The transfer of those vectors toward the GPU memory could slow down the GPU performance. I’ve decided to read the ionization tables and calculate the corresponding quantities (alpha, beta, gamma, lambda) directly in the ionization rate function, without using those vector members. Of course, I still need to do the performance benchmark of my solution against the master branch to determine which is the best.

GianlucaOberreit commented:
Thank you!

> Are you planning to make the ionisation rate also inherit from the ionisation tunnel?

I wanted to do this but unfortunately don't currently see an elegant way of achieving this whilst maintaining backwards compatibility.

For the rest, sounds good!

I haven't done any performance checks for my modifications yet, so I don't know if some of my decisions had a negative impact.
